File: leetcode/cyclic_sort/268_missing_number.py
'''
LeetCode 268 - Missing Number [easy]

Given an array containing n distinct numbers taken from 0, 1, 2, ..., n,
find the one that is missing from the array.

Example 1:

Input: [3, 0, 1]
Output: 2
Example 2:

Input: [9, 6, 4, 2, 3, 5, 7, 0, 1]
Output: 8
Note:

Your algorithm should run in linear runtime complexity. Could you implement it using only constant extra space complexity?
'''
from typing import List


class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        '''
        - cyclic sort, look for range of numbers and find the missing
        - this is only truly efficient if the numbers can be used as indexes otherwise a cycle sort is O(n**2) since we have to scan each pass through
        - if the item is not in position swap it with the index spot, otherwise ignore
        - when everything is in place loop through and find the index of the biggest item
        '''
        start = 0
        hi = len(nums)
        while start < hi:
            num = nums[start]
            # compare the index and the num
            if num == start or num == hi:
                start += 1
            else:
                nums[num], nums[start] = nums[start], nums[num]

        for i in range(hi):
            if nums[i] == hi:
                return i

        # if everything is in place then the big number is missing
        return hi


sol = Solution()
print(sol.missingNumber([3, 0, 1]))

File: js/cyclic_sort/268_missing_number.js
/**
 * 
 * Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.

Follow up: Could you implement a solution using only O(1) extra space complexity and O(n) runtime complexity?



Example 1:

Input: nums = [3,0,1]
Output: 2
Explanation: n = 3 since there are 3 numbers, so all numbers are in the range [0,3]. 2 is the missing number in the range since it does not appear in nums.
Example 2:

Input: nums = [0,1]
Output: 2
Explanation: n = 2 since there are 2 numbers, so all numbers are in the range [0,2]. 2 is the missing number in the range since it does not appear in nums.
Example 3:

Input: nums = [9,6,4,2,3,5,7,0,1]
Output: 8
Explanation: n = 9 since there are 9 numbers, so all numbers are in the range [0,9]. 8 is the missing number in the range since it does not appear in nums.
Example 4:

Input: nums = [0]
Output: 1
Explanation: n = 1 since there is 1 number, so all numbers are in the range [0,1]. 1 is the missing number in the range since it does not appear in nums.
 */

/**
 * @param {number[]} nums
 * @return {number}
 * 
 * if i == ival do nothing
 * if i is max number do nothing
 * swap i val with i
 * 
 * if see n do nothing or if i === val do nothing
 * else swap with i
 */
var missingNumber = function (nums) {
  let i = 0

  while (i < nums.length) {
    const val = nums[i]
    if (i === val || val === nums.length) {
      i++
    } else {
      // why not go forward on swap?
      // because what you just swapped is not in place yet
      // for loop would have skipped this
      nums[i] = nums[val]
      nums[val] = val
    }
  }

  let out = nums.findIndex(item => item === nums.length)

  return out !== -1 ? out : nums.length
};


console.log(missingNumber([3, 0, 1]))
console.log(missingNumber([9, 6, 4, 2, 3, 5, 7, 0, 1]))
